yapapi/contrib/service/http_proxy.py

```diff
@@ -0,0 +1,173 @@
+        self.headers_data = b""
+        self.headers: Optional[CIMultiDict] = None
+        self.content = b""
+        self.status: Optional[int] = None
+        self.ws = ws
+        self.timeout = timeout
+
+    def process_headers(self):
+        header_lines = self.headers_data.decode("ascii").splitlines()
+
+        self.status = int(header_lines[0].split()[1])
+        self.headers = CIMultiDict()
+
+        for line in header_lines[1:]:
+            if line:
+                name, value = line.split(": ", maxsplit=1)
+                self.headers[name] = value
+
+    def receive_data(self, data: bytes):
+        if self.status:
+            self.content += data
+        else:
+            parts = re.split(b"\r\n\r\n", data)
+            self.headers_data += parts.pop(0)
+            if parts:
+                self.content += parts.pop(0)
+                self.process_headers()
+
+    @property
+    def content_received(self) -> bool:
+        if not self.status:
+            return False
+
+        assert self.headers is not None
+        content_length = self.headers.get("Content-Length")
+
+        if not content_length:
+            return True
+
+        return len(self.content) >= int(content_length)
+
+    async def get_response(self) -> web.Response:
+        while not self.content_received:
+            l = await self.ws.receive(self.timeout)
+            self.receive_data(l.data)
+
+        assert self.status
+        return web.Response(status=self.status, headers=self.headers, body=self.content)
+
+
+class HttpProxyService(Service, abc.ABC):
+    def __init__(
+        self,
+        remote_port: int = 80,
+        remote_host: Optional[str] = None,
+        response_timeout: float = 10.0,
+    ):
+        super().__init__()
+        self._remote_port = remote_port
+        self._remote_host = remote_host
+        self._remote_response_timeout = response_timeout
+
+    async def handle_request(self, request: web.Request) -> web.Response:
+        """
+        handle the request coming from the local HTTP server
+        by passing it to the instance through the VPN
+        """
+        assert self.network_node, "Ensure that the service is started within a network."
+        assert self.cluster
+
+        instance_ws = self.network_node.get_websocket_uri(self._remote_port)
+        app_key = self.cluster.service_runner._job.engine._api_config.app_key  # noqa
+
+        remote_headers = "\r\n".join(
+            [
+                f"{k}: {v if k != 'Host' else self._remote_host or v}"
+                for k, v in request.headers.items()
+            ]
+        )
+
+        remote_request: bytes = (
+            f"{request.method} {request.path_qs} "
+            f"HTTP/{request.version.major}.{request.version.minor}\r\n"
+            f"{remote_headers}\r\n\r\n"
+        ).encode("ascii") + (await request.read() if request.can_read_body else b"")
```

Comment:
Also I still think this is super ugly and unreadable at all. I just don't understand why we want to have such things.

Reply:
addressed